File: CS/WindowsApplication1/CustomControl.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Text;
using System.Windows.Forms;
using DevExpress.Utils;

namespace WindowsApplication1
{
    public class CustomControl : Control, IToolTipControlClient
    {

        public CustomControl()
        {

        }

        protected override void OnPaint(PaintEventArgs e)
        {
            base.OnPaint(e);
            MyDrawControl(e);
        }
        private void MyDrawControl(PaintEventArgs e)
        {
            DrawBackGround(e);
            DrawLine(e);
            DrawStart(e);
            DrawFinish(e);
        }

        private void DrawBackGround(PaintEventArgs e)
        {
            e.Graphics.FillRectangle(new SolidBrush(BackColor), ClientRectangle);
        }
        private void DrawLine(PaintEventArgs e)
        {
            e.Graphics.FillRectangle(new SolidBrush(ForeColor), GetLineBounds(e.Graphics));
        }

        private void DrawStart(PaintEventArgs e)
        {
            Rectangle rect = GetStartBounds(e.Graphics);
            e.Graphics.DrawString("Start", Font, new SolidBrush(ForeColor), rect);
        }

        private void DrawFinish(PaintEventArgs e)
        {
            Rectangle rect = GetFinishBounds(e.Graphics);
            e.Graphics.DrawString("Finish", Font, new SolidBrush(ForeColor), rect);
        }

        private RectangleF GetLineBounds(Graphics e)
        {
            Rectangle rect1 = GetStartBounds(e);
            Rectangle rect2 = GetFinishBounds(e);
            int lineWidth = Width - rect1.Width - rect2.Width;
            Rectangle rect = new Rectangle(rect1.Right, 0, lineWidth, Height);
            return rect;
        }


        private Rectangle GetStartBounds(Graphics g)
        {
            Size size = g.MeasureString("Start", Font).ToSize();
            return new Rectangle(0, 0, size.Width + 5, Height);
        }

        private Rectangle GetFinishBounds(Graphics g)
        {
            Size size = g.MeasureString("Finish", Font).ToSize();
            return new Rectangle(Width - size.Width, 0, size.Width + 5, Height);
        }

        public string GetHintText(Point p)
        {
            Graphics graphics = CreateGraphics();
            if (GetStartBounds(graphics).Contains(p))
                return "Start";
            if (GetFinishBounds(graphics).Contains(p))
                return "Finish";
            if (GetLineBounds(graphics).Contains(p))
                return "Line";
            return "Custom control";
        }

        #region IToolTipControlClient Members

        ToolTipControlInfo IToolTipControlClient.GetObjectInfo(Point point)
        {
            string hintText = GetHintText(point);
            ToolTipControlInfo info = new ToolTipControlInfo();
            info.Object = hintText;
            info.Text = hintText;
            info.ToolTipPosition = Control.MousePosition;
            info.Interval = 100;
            info.Title = "Custom title";
            info.ToolTipType = ToolTipType.SuperTip;
            return info;
        }



        bool IToolTipControlClient.ShowToolTips
        {
            get { return true; }
        }

        #endregion
    }
}

File: VB/WindowsApplication1/CustomControl.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Data
Imports System.Drawing
Imports System.Text
Imports System.Windows.Forms
Imports DevExpress.Utils

Namespace WindowsApplication1
	Public Class CustomControl
		Inherits Control
		Implements IToolTipControlClient

		Public Sub New()

		End Sub

		Protected Overrides Sub OnPaint(ByVal e As PaintEventArgs)
			MyBase.OnPaint(e)
			MyDrawControl(e)
		End Sub
		Private Sub MyDrawControl(ByVal e As PaintEventArgs)
			DrawBackGround(e)
			DrawLine(e)
			DrawStart(e)
			DrawFinish(e)
		End Sub

		Private Sub DrawBackGround(ByVal e As PaintEventArgs)
			e.Graphics.FillRectangle(New SolidBrush(BackColor), ClientRectangle)
		End Sub
		Private Sub DrawLine(ByVal e As PaintEventArgs)
			e.Graphics.FillRectangle(New SolidBrush(ForeColor), GetLineBounds(e.Graphics))
		End Sub

		Private Sub DrawStart(ByVal e As PaintEventArgs)
			Dim rect As Rectangle = GetStartBounds(e.Graphics)
			e.Graphics.DrawString("Start", Font, New SolidBrush(ForeColor), rect)
		End Sub

		Private Sub DrawFinish(ByVal e As PaintEventArgs)
			Dim rect As Rectangle = GetFinishBounds(e.Graphics)
			e.Graphics.DrawString("Finish", Font, New SolidBrush(ForeColor), rect)
		End Sub

		Private Function GetLineBounds(ByVal e As Graphics) As RectangleF
			Dim rect1 As Rectangle = GetStartBounds(e)
			Dim rect2 As Rectangle = GetFinishBounds(e)
			Dim lineWidth As Integer = Width - rect1.Width - rect2.Width
			Dim rect As New Rectangle(rect1.Right, 0, lineWidth, Height)
			Return rect
		End Function


		Private Function GetStartBounds(ByVal g As Graphics) As Rectangle
			Dim size As Size = g.MeasureString("Start", Font).ToSize()
			Return New Rectangle(0, 0, size.Width + 5, Height)
		End Function

		Private Function GetFinishBounds(ByVal g As Graphics) As Rectangle
			Dim size As Size = g.MeasureString("Finish", Font).ToSize()
			Return New Rectangle(Width - size.Width, 0, size.Width + 5, Height)
		End Function

		Public Function GetHintText(ByVal p As Point) As String
			Dim graphics As Graphics = CreateGraphics()
			If GetStartBounds(graphics).Contains(p) Then
				Return "Start"
			End If
			If GetFinishBounds(graphics).Contains(p) Then
				Return "Finish"
			End If
			If GetLineBounds(graphics).Contains(p) Then
				Return "Line"
			End If
			Return "Custom control"
		End Function

		#Region "IToolTipControlClient Members"

		Private Function GetObjectInfo(ByVal point As Point) As ToolTipControlInfo Implements IToolTipControlClient.GetObjectInfo
			Dim hintText As String = GetHintText(point)
			Dim info As New ToolTipControlInfo()
			info.Object = hintText
			info.Text = hintText
			info.ToolTipPosition = Control.MousePosition
			info.Interval = 100
			info.Title = "Custom title"
			info.ToolTipType = ToolTipType.SuperTip
			Return info
		End Function



		Private ReadOnly Property ShowToolTips() As Boolean Implements IToolTipControlClient.ShowToolTips
			Get
				Return True
			End Get
		End Property

		#End Region
	End Class
End Namespace
